### PR TITLE
refactor(array): use `into` for append datum matching [IGNORE ME]

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

As suggested at https://github.com/risingwavelabs/risingwave/issues/9113#issuecomment-1527018228, this reduces the domain of `match` and may improve the performance (?).

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/common/src/array/mod.rs [hunk 1]
```diff
@@ -483,14 +483,9 @@ macro_rules! impl_array_builder {
...
-                    Some(scalar_ref) => match (self, scalar_ref) {
-                        $( (Self::$variant_name(inner), ScalarRefImpl::$variant_name(v)) => inner.append_n(n, Some(v)), )*
-                        (this_builder, this_scalar_ref) => panic!(
-                            "Failed to append datum, array builder type: {}, scalar type: {}",
-                            this_builder.get_ident(),
-                            this_scalar_ref.get_ident()
-                        ),
-                    },
+                    Some(scalar_ref) => paste::paste! { match self {
+                        $( Self::$variant_name(inner) => inner.append_n(n, Some(scalar_ref.[<into_ $suffix_name>]())), )*
```
Comment (kwannoel) [inline src/common/src/array/mod.rs:487]: Does `into` also result in `n` lookup?

Comment (BugenZhao) [reply src/common/src/array/mod.rs:487]: I'm not sure about the compiler's optimization. 🥺 Let's wait for the flamegraph.

**Comment by lmatz** [comment]
SCR-20230428-kmy

**Comment by kwannoel** [comment]
Today's `main-cron`:

Screenshot 2023-04-28 at 2 52 46 PM

**Comment by BugenZhao** [comment]
Seems no obvious improvements. 😟

**Comment by lmatz** [comment]
It seems inevitable to touch `SourceStreamChunkRowWriter` 🥵

**Comment by kwannoel** [comment]
Maybe can try `union` also? 

```
dyn_no_info             time:   [198.01 ns 198.49 ns 198.98 ns]
                        change: [-1.2664% -0.7003% -0.1354%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

union dyn               time:   [150.13 ns 150.60 ns 151.08 ns]
                        change: [-0.0452% +0.7195% +1.4891%] (p = 0.08 > 0.05)
                        No change in performance detected.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) high mild
  1 (1.00%) high severe
```

Did an experiment with `union` comparing these:

https://github.com/kwannoel/rust-experiments/blob/fb7a604bc7ab364ca21a60a1ea808dcf6779dab1/benches/match-dyn.rs#L207-L214
```
    fn append_u(&mut self, d: DynE, x: DynU) {
         match d {
           DynE::A => { unsafe { self.a.append_u(x) } },
           DynE::B => { unsafe { self.b.append_u(x) } },
           DynE::C => { unsafe { self.c.append_u(x) } },
         }
    }
```

https://github.com/kwannoel/rust-experiments/blob/fb7a604bc7ab364ca21a60a1ea808dcf6779dab1/benches/match-dyn.rs#L137-L144
```
    fn append_no_info(&mut self, x: Dyn) {
        match (self, x) {
            (Dyns::A(xs), Dyn::A(x)) => { xs.append(x) },
            (Dyns::B(xs), Dyn::B(x)) => { xs.append(x) },
            (Dyns::C(xs), Dyn::C(x)) => { xs.append(x) },
            (_, _) => { panic!("failed...") }
        }
    }
```

Seems like with `union` (first one), we can get 25% improvement. Trade-off is using unsafe code, and perhaps making the code more complex.

**Comment by BugenZhao** [comment]
> Maybe can try `union` also?

It seems that `std::hint::unreachable_unchecked` on enum matching has the same effect as `union`. 🤔 I'd prefer to avoid introducing unsafe structures for this.

[Comparison](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DEArgoKkl9ZATwDKjdAGFUtEywYTpjgDJ4GmABy7gBGmMQgAEzSAA6oCoR2DC5uHl6kcQm2An4BwSxhEdGWmNbZDEIETMQEKe6eUiVlSZXVBLlBoeFR0uZtdWmNfTUd%2BYU9AJSWqCbEyOwcZpgA1OboICAI/gQbIQbIANYA%2BiGoqgCkAMwAQucaAIJ394zuywAiTFUAKgCeMSvnADstweyzBywAkpdIqQnuDlgAxABs0ieQLeaIeJgYSWWchEYmqEIYAWIyyBIPu8Lw0JAyxpMLh4KoKLpLNRD3RT3MphseIJBjJFKZYIIPzpH2%2Bf0wsNB4P8pLp%2BNEguJpNlj0BGIeeBYMVo/JV1XJwJFy2xCiYVBWVAY9MEqCODKO2OQCEwh0w6AgVioE2WAForth6dCTZT4fDfQA6BXhWPQs1czlazHPBivbAC43CuVgqGRCAMqZm5GSagoksp7X3XX65ZZo1C01580MS3W5a2%2B1EJ3Ql0MN0eg5en2lP2B4OhyLhs3wlifN2rcez1uR8GNwkREAFov%2Bq4Yy4hvAa9fro7ky6HkNrDZbQQbbHETBMN1MPaYAdDz3eqtUs/Jv%2BYKASBnKRJc5wAKzXP4tD%2BJgEABAAbuEExQTW3YmD6SrZsQ/pBke06rkBbYdgCwKrLGDp9pEX7uj%2BED7qm1ZouBUEwQwcEBIhmAoXh6FPN2CEKHSm6CvhU4MsR8IWla5HXJR2yOs6rr0SOv4mjWoH3DEJghF2doLv4jHSeCeyvscpyqBAWHmYcJxnBAypbqZ65ihKnxML8/wbAWp5nnGESGluarhOG050uBlyaf5wFahMCVXBGZn7JZjkIXZaXWWJ1S%2BdCEBRQliU3GizH3BwUy0JwkG8J4HBaKQqCcAASmYBCrDMcwAuBPCkDs9UVVMBwgJBGj6Jwki1ZovBNRwvAiWN/VaFMcCwEgaB6nQ4TkJQG0xFtEQ4sACAELQPykFgSF4PMABqeCYAA7gA8v8dW9TQtAEOEIkQCE02kCE/jVOK3C8IDzDED8T0hNoHr9b1G1sIIT2cSDDVYAuRjiANF14M%2BNh4ChIk45gqgeiYX2cL12ylP9cEhMQwMuFg/0EMQupUxVfAGMACh3Y9L2MJzMiCAK7CNPwgiKCo6g47oXD6IYxhtfoeAhCJkBTKgMTlMTAZPfNpRw0kDgMM4rj1HovjwWM3QK5kiQCAMDQZPEjsMKMXQRArVjGwIrQ1M7ei%2BwT/tVCMNte8H4e1BbgyWDHnsFHbUwKJ18wSJV1VTTjs3LKY5jIMsR0nWdywQLghAkOSPUTLwS2DaQw2jeNHCTaQdUNbN80gIt00Ja3kQ513nD1/3Ux8Qk9iSEAA%3D%3D%3D)

**Pull request closed without merging**